src/styles/text/text.js

```diff
@@ -0,0 +1,377 @@
+
+                for (let text in text_infos) {
+                    let text_info = text_infos[text];
+                    let label;
+                    let keep_in_tile;
+                    let move_in_tile;
+                    let geometry = this.geometries[tile][style][text];
+                    let exceed_heuristic = this.label_style.lines.exceed;
+
+                    if (geometry.type === "LineString") {
+                        let lines = geometry.coordinates;
+                        let line = [lines[0]];
+
+                        keep_in_tile = true;
+                        move_in_tile = true;
+
+                        label = new Label(text, line[0], text_info.size, lines);
+                    } else if (geometry.type === "Point") {
+                        let points = [geometry.coordinates];
+
+                        keep_in_tile = true;
+                        move_in_tile = false;
+
+                        label = new Label(text, points[0], text_info.size);
+                    }
+
+                    if (label.discard(move_in_tile, keep_in_tile, this.bboxes[tile], exceed_heuristic)) {
+                        // remove the text from the map
+                        delete text_infos[text];
+                    }
+
+                    text_info.label = label;
+                }
+
+                // No labels for this style
+                if (Object.keys(text_infos).length === 0) {
+                    delete texts[style];
+                }
+            }
+
+            // No labels for this tile
+            if (Object.keys(texts).length === 0) {
+                WorkerBroker.postMessage('TextStyle', 'freeTile', tile);
+                // early exit
+                return;
+            }
+
+            // second call to main thread, for rasterizing the set of texts
+            return WorkerBroker.postMessage('TextStyle', 'addTexts', tile, texts).then(({ texts, texture }) => {
+                this.texts[tile] = texts;
+
+                // Attach tile-specific label atlas to mesh as a texture uniform
+                tile_data.uniforms = { u_textures: [texture] };
+
+                // Build queued features
+                tile_data.queue.forEach(q => this.super.addFeature.apply(this, q));
+                tile_data.queue = [];
+                return this.super.endData.call(this, tile_data);
+            });
+        });
+    },
+
+    // Override to queue features instead of processing immediately
+    addFeature (feature, rule, context, tile_data) {
+        // Collect text
+        let text = feature.properties.name;
+
+        if (text) {
+            let tile = context.tile.key;
+            if (!this.texts[tile]) {
+                this.texts[tile] = {};
+            }
+
+            if (!this.geometries[tile]) {
+                this.geometries[tile] = {};
+            }
+
+            let style = this.constructFontStyle(rule);
+            let style_key = this.constructStyleKey(style);
```

Comment:
This could get messy as we add more style params, but might be the only reasonable solution anyway.